src/btree/bt_curnext.c

```diff
@@ -402,6 +403,17 @@ __cursor_row_next(WT_CURSOR_BTREE *cbt, bool newpage, bool restart, size_t *skip
 restart_read_page:
         rip = &page->pg_row[cbt->slot];
         WT_RET(__cursor_row_slot_key_return(cbt, rip, &kpack, &kpack_used));
+        /*
+         * If the cursor has prefix search configured we can early exit here, given the supplied
+         * prefix doesn't match the key we have.
+         */
+        if (F_ISSET(&cbt->iface, WT_CURSTD_PREFIX_SEARCH) && prefix != NULL &&
+          !__wt_prefix_match(prefix, &cbt->iface.key)) {
+            /* It is not okay for the user to have a custom collator. */
+            WT_ASSERT(session, CUR2BT(cbt)->collator == NULL);
+            WT_STAT_CONN_DATA_INCR(session, cursor_search_near_prefix_fast_paths);
+            return (WT_NOTFOUND);
```

Comment:
I'm still concerned that this code is returning WT_NOTFOUND without resetting the cursor state. I *think* it's necessary so that the subsequent prev operation can work as expected, but it's a surprising semantic, and one we would want to be different if (for example) we introduced prefixes as a way to constrain range scans with cursors in the future - so it deserves a big comment describing why it's OK (i.e: that the only caller that can end up taking this path has special handling of WT_NOTFOUND, which involves issuing a subsequent operation on the cursor, so expects the cursor to be positioned even though not found is being returned.

Reply:
I left a comment on the caller to this as felt like that was the place that that should be explained as it's the caller that does the reset.

Reply:
I see now. Makes sense.